### PR TITLE
Move CFFI to Cython (+address reviews)

### LINKED ISSUE #40: [FEA] Replace CFFI bindings with Cython
We should transition the current Python bindings API to Cython in order
 - to support promoting C++ exceptions up to Python exceptions for better error reporting
 - to match the approach cuDF is taking

### PR DESCRIPTION
**Pull request by shwina**
Addresses review comments in #132 

Fixes #40

### CODE REVIEW

**Comment by kkraus14** [comment]
@shwina should just change the PR # in the changelog to reflect this PR

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@
 ## Improvements
  - PR #123 Remove driver install from ci scripts
  - PR #131 Use YYMMDD tag in nightly build
+ - PR #137 Move CFFI to Cython
```
Comment (harrism) [inline CHANGELOG.md]: Can we make this more informative? Like
```suggestion
 - PR #137 Replace CFFI python bindings with Cython
```

File: python/.coveragerc [hunk 2]
```diff
@@ -0,0 +1,4 @@
+# Configuration file for Python coverage tests
+[run]
+include = rmm/*
+omit = rmm/tests/*
```
Comment (harrism) [inline python/.coveragerc]:
```suggestion
omit = rmm/tests/*

```

File: python/rmm/_lib/lib.pxd [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+    cdef size_t rmmLogSize() except +
+
+    cdef rmmError_t rmmGetLog(
+        char* buffer,
+        size_t buffer_size
+    ) except +
+
+
+cdef extern from "rmm/rmm.hpp" namespace "rmm" nogil:
```
Comment (harrism) [inline python/rmm/_lib/lib.pxd]: Note that I think eventually we may eliminate the C interface above and everything will move to the C++ API. CFFI was blocking this.

File: python/rmm/_lib/lib.pyx [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+def rmm_alloc(size, stream):
+    """
+    Allocates size bytes using the RMM memory manager by calling the librmm
+    functions via Cython
+    """
+    cdef size_t c_size = size
+    cdef cudaStream_t c_stream = <cudaStream_t><size_t>stream
+
+    cdef uintptr_t c_addr = c_alloc(
```
Comment (harrism) [inline python/rmm/_lib/lib.pyx]: Can you explain what the extra `c_alloc` layer is?

File: setup.cfg [hunk 5]
```diff
@@ -0,0 +1,7 @@
+[flake8]
```
Comment (harrism) [inline setup.cfg]: This is in the root directory? Does it belong under the python directory?

File: python/rmm/_lib/lib.pyx (hunk 4 above)
Comment (kkraus14) [reply python/rmm/_lib/lib.pyx]: This is a Cython API which is also exposed in the `lib.pxd` file, which allows another Cython project to depend on the `rmm` Python package and use these functions from within Cython. I though about getting rid of the `c_*` layer and just having the RMM C/C++ API and the Python API via Cython, but it felt like there was a bit of an awkward gap for Cython.

File: setup.cfg (hunk 5 above)
Comment (kkraus14) [reply setup.cfg]: There's one in the python directory already, I vaguely remember there being a reason for adding this but let me try removing it and report back.

**Comment by kkraus14** [comment]
@harrism addressed the changes in your review, could you give it another look when you get the chance?

File: python/rmm/_lib/lib.pxd (hunk 3 above)
Comment (kkraus14) [reply python/rmm/_lib/lib.pxd]: Yup understood, will adjust the Cython in lockstep with those changes, didn't want to rock the boat for now though 😄

Comment (kkraus14) [reply python/rmm/_lib/lib.pxd]: Yup understood, will adjust the Cython in lockstep with those changes, didn't want to rock the boat for now though 😄

**Pull request merged**